pkg/controller/workerroute/workerroute_controller.go

```diff
@@ -0,0 +1,291 @@
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &WorkerRouteReconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("workerroute-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to WorkerRoute
+	err = c.Watch(&source.Kind{
+		Type: &crdsv1alpha1.WorkerRoute{},
+	}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return errors.Wrap(err, "failed to start watch on workerroutes")
+	}
+
+	generatedClient := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+	generatedInformers := kubeinformers.NewSharedInformerFactory(generatedClient, time.Minute)
+	err = mgr.Add(manager.RunnableFunc(func(s <-chan struct{}) error {
+		generatedInformers.Start(s)
+		<-s
+		return nil
+	}))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WorkerRouteReconciler reconciles a WorkerRoute object
+type WorkerRouteReconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=crds.kubeflare.io,resources=workerroutes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=crds.kubeflare.io,resources=workerroutes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=crds.kubeflare.io,resources=workerroutes/finalizers,verbs=update
+func (r *WorkerRouteReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+
+	var instance crdsv1alpha1.WorkerRoute
+	if err := r.Get(ctx, request.NamespacedName, &instance); err != nil {
+		if !apiErrors.IsNotFound(err) {
+			logger.Error(err)
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	zone, err := shared.GetZone(ctx, instance.Namespace, instance.Spec.Zone)
+	if err != nil {
+		logger.Error(err)
+		return reconcile.Result{}, ignoreUnrecoverableErrors(err)
+	}
+
+	cf, err := shared.GetCloudflareAPI(ctx, instance.Namespace, zone.Spec.APIToken)
+	if err != nil {
+		logger.Error(err)
+		return reconcile.Result{}, ignoreUnrecoverableErrors(err)
+	}
+
+	zoneID, err := cf.ZoneIDByName(zone.Name)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to get zone id")
+	}
+
+	if err := r.ReconcileWorkerRouteInstances(ctx, &instance, zoneID, cf); err != nil {
+		logger.Error(err)
+		return reconcile.Result{}, ignoreUnrecoverableErrors(err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *WorkerRouteReconciler) ReconcileWorkerRouteInstances(ctx context.Context, instance *crdsv1alpha1.WorkerRoute, zoneID string, cf *cloudflare.API) error {
+	logger.Debug("reconcile worker route", zap.String("name", instance.Name))
+
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
I think we can simplify this if/else block.
It would also be better to use `controllerutil.ContainsFinalizer` so we won't need to define `ContainsString`